src/Datadog.Trace/Propagators/HttpHeadersPropagator.cs

```diff
@@ -1,32 +1,46 @@
-﻿using System.Net.Http.Headers;
+using System.Globalization;
+using System.Linq;
+using System.Net.Http.Headers;
+using Datadog.Trace.ExtensionMethods;
 
 namespace Datadog.Trace.Propagators
 {
-    /// <summary>
-    /// A HTTP Propagator using <see cref="HttpHeaders"/> as carrier
-    /// </summary>
-    public static class HttpHeadersPropagator
+    internal class HttpHeadersPropagator : IPropagator
     {
-        /// <summary>
-        /// Extracts a <see cref="SpanContext"/> from the specified headers.
-        /// </summary>
-        /// <param name="headers">The headers.</param>
-        /// <returns>The <see cref="SpanContext"/></returns>
-        public static SpanContext Extract(this HttpHeaders headers)
+        private readonly HttpHeaders _headers;
+
+        public HttpHeadersPropagator(HttpHeaders headers)
+        {
+            _headers = headers;
+        }
+
+        public void Inject(SpanContext context)
         {
-            var wrapper = new HttpHeadersWrapper(headers);
-            return wrapper.Extract();
+            _headers.Remove(HttpHeaderNames.TraceId);
+            _headers.Add(HttpHeaderNames.TraceId, context.TraceId.ToString(CultureInfo.InvariantCulture));
+
+            _headers.Remove(HttpHeaderNames.ParentId);
+            _headers.Add(HttpHeaderNames.ParentId, context.SpanId.ToString(CultureInfo.InvariantCulture));
         }
 
-        /// <summary>
-        /// Injects the specified <see cref="SpanContext"/> in the <see cref="HttpHeaders"/>.
-        /// </summary>
-        /// <param name="headers">The headers.</param>
-        /// <param name="context">The context.</param>
-        public static void Inject(this HttpHeaders headers, SpanContext context)
+        public SpanContext Extract()
         {
-            var wrapper = new HttpHeadersWrapper(headers);
-            wrapper.Inject(context);
+            ulong? traceId = null;
+            ulong? parentId = null;
+
+            if (_headers.TryGetValues(HttpHeaderNames.TraceId, out var traceIds))
```

Comment:
Seems this was removed in httpclient pr, not sure if it's staying or going, but if it's staying the logic comments about getting the first parsed vs. first then parse would apply here if staying